Add workaround for invalid buffering info on android

timoxd7 opened the pull request:
## Info
As already described in multiple issues, there is currently a problem with android reporting "isBuffering" with true even if the video is already playing, preventing the video controls to be displayed. This is a workaround, checking (on android only) if the video is actually buffering. Tested with a network- and file-based video.

>NOTE: This is only a workaround until video_player and flutter is fixed. See https://github.com/flutter/flutter/issues/165149 for more information. Also thanks to @wenwangwill who added a workaround as comment i got inspired from.

## Additional Issues not solved
If pausing the video and seeking to some point (using seekTo, so clicking on the progress bar) still leads to the infinit loading bar, as no callback is called by the video controller and thus the buffering not updated.

As workaround, just start playing on seekTo like this:

```dart
  // Your init code can be above
  videoController.addListener(yourListeningMethod);

  // ...

  bool wasPlayingBefore = false;
  void yourListeningMethod() {
    if (!videoController.value.isPlaying && !wasPlayingBefore) {
      // -> Workaround if seekTo another position while it was paused before
      //    on Android this might lead to infinite loading, so just play the
      //    video again.
      videoController.play();
    }

    wasPlayingBefore = videoController.value.isPlaying;

  // ...
  }

```

## Related to
#909
#907 
#906
#903

diegotori commented on lib/src/cupertino/cupertino_controls.dart:
```diff
@@ -810,9 +811,37 @@ class _CupertinoControlsState extends State<CupertinoControls>
 
+    late final bool buffering;
+
+    if (Platform.isAndroid) {
+      if (controller.value.isBuffering) {
+        // -> Check if we actually buffer, as android has a bug preventing to
+        //    get the correct buffering state from this single bool.
+        final VideoPlayerValue value = controller.value;
+
+        final int position = value.position.inMilliseconds;
+
+        // Special case, if the video is finished, we don't want to show the
+        // buffering indicator anymore
+        if (position >= value.duration.inMilliseconds) {
+          buffering = false;
+        } else {
+          final int buffer =
+              value.buffered.lastOrNull?.end.inMilliseconds ?? -1;
+
+          buffering = position >= buffer;
+        }
+      } else {
+        // -> No buffering
+        buffering = false;
+      }
+    } else {
+      buffering = controller.value.isBuffering;
+    }
```
Encapsulate this into a private function.

diegotori commented on lib/src/material/material_controls.dart:
```diff
@@ -645,9 +646,37 @@ class _MaterialControlsState extends State<MaterialControls>
 
+    late final bool buffering;
+
+    if (Platform.isAndroid) {
+      if (controller.value.isBuffering) {
+        // -> Check if we actually buffer, as android has a bug preventing to
+        //    get the correct buffering state from this single bool.
+        final VideoPlayerValue value = controller.value;
+
+        final int position = value.position.inMilliseconds;
+
+        // Special case, if the video is finished, we don't want to show the
+        // buffering indicator anymore
+        if (position >= value.duration.inMilliseconds) {
+          buffering = false;
+        } else {
+          final int buffer =
+              value.buffered.lastOrNull?.end.inMilliseconds ?? -1;
+
+          buffering = position >= buffer;
+        }
+      } else {
+        // -> No buffering
+        buffering = false;
+      }
+    } else {
+      buffering = controller.value.isBuffering;
+    }
```
Encapsulate this into a private function.

diegotori commented on lib/src/material/material_desktop_controls.dart:
```diff
@@ -581,9 +582,37 @@ class _MaterialDesktopControlsState extends State<MaterialDesktopControls>
 
+    late final bool buffering;
+
+    if (Platform.isAndroid) {
+      if (controller.value.isBuffering) {
+        // -> Check if we actually buffer, as android has a bug preventing to
+        //    get the correct buffering state from this single bool.
+        final VideoPlayerValue value = controller.value;
+
+        final int position = value.position.inMilliseconds;
+
+        // Special case, if the video is finished, we don't want to show the
+        // buffering indicator anymore
+        if (position >= value.duration.inMilliseconds) {
+          buffering = false;
+        } else {
+          final int buffer =
+              value.buffered.lastOrNull?.end.inMilliseconds ?? -1;
+
+          buffering = position >= buffer;
+        }
+      } else {
+        // -> No buffering
+        buffering = false;
+      }
+    } else {
+      buffering = controller.value.isBuffering;
+    }
```
Encapsulate this into a private function.

yann-elpyoo commented:
For information until the bug is fixed, **and for chewie users only** : it is possible to use a simple and complete workaround.

The thing is to add a long duration for the loader to appear. Then nothing is blocked anymore even at the end of the video! No need for looping too. It just shows the video paused and played each time seek to is called... 
```
 _chewieController = ChewieController(
          videoPlayerController: _videoPlayerController,
          progressIndicatorDelay:
              PlatformProvider.isAndroid() ? const Duration(days: 1) : null,
        );
```

For non chewie users suffering from it, maybe wrapping with chewie could be a temporary solution...

naghinezhad commented:
@timoxd7 @diegotori
Please review and merge this pull request so the related bug can be resolved. We are waiting for it.

timoxd7 commented:
@naghinezhad i am a bit in two minds about this pr. I will incorporate the requested changes, but it still is not a complete fix. 

I don't know if it works good with live streams. As @yann-elpyoo wrote, you _could_ fix it by disabling indirectly the loading spinner for android, but this also removes the feedback to the user that something is buffering.

@diegotori i updated according to your review. I also tested @yann-elpyoo solution and added it to the readme, both work together too (as my workaround will just be _hidden_ by disabling the loader anyway).

As maintainer you can decide wether to merge this, as the root problem is not relying in chewie itself. However, we now have it documented here at least and if someone needs a fix, can use it util it is fixed in flutter itself.

diegotori commented on lib/src/helpers/utils.dart:
```diff
@@ -1,3 +1,7 @@
+import 'dart:io';
```
Instead, have that method take in a `BuildContext` and then use `(Theme.of(context).platform` to find the current platform.

timoxd7 replied:
Will fix it!

timoxd7 replied:
Just saw your second comment. Should i change it still to `BuildContext`?

diegotori reviewed (changes requested):
For the new utility method, add comments related to the issue your changes are trying to address. 

Thanks.

diegotori commented on lib/src/helpers/utils.dart:
```diff
@@ -34,6 +34,11 @@ String formatDuration(Duration position) {
   return formattedTime;
 }
 
+/// Get the current buffering state of the video player. Will invoke a
+/// Workaround for Android, as a bug in the video_player plugin prevents to get
+/// the actual buffering state, always returning true and breaking ui elements.
+/// For this, the actual buffer position is used to determine if the video is
+/// buffering or not. See #912 for more details.
```
Change this to:

```dart
/// Gets the current buffering state of the video player. 
/// 
/// For Android, it will use a workaround due to a [bug](https://github.com/flutter/flutter/issues/165149) 
/// affecting the `video_player` plugin, preventing it from getting the 
/// actual buffering state. This currently results in the `VideoPlayerController` always buffering,
/// thus breaking UI elements.
///
/// For this, the actual buffer position is used to determine if the video is
/// buffering or not. See Issue [#912](https://github.com/fluttercommunity/chewie/pull/912) for more details.
```

diegotori commented:
@timoxd7 other than the final nit that I called out, the rest LGTM.

diegotori reviewed (approved):
LGTM.
